### ISSUE TITLE
Why Idenity is not used during verifier calculation?

### ISSUE THREAD

**Issue by JexSrs** [midonet/tssrp6a#87]
Why Identity is not used during verifier calculation? According to some posts like this [one](https://crypto.stackexchange.com/a/9430/80987) it is a big security flaw. Also [RFC 5054](https://www.rfc-editor.org/rfc/rfc5054#section-2.4) uses identity during hashing. Is there a  reason why the user's identity (username/email) is skipped?

That was found [here](https://github.com/midonet/tssrp6a/blob/ec90b27105b2b18cb7b57d469ffa2d5194bde19f/src/routines.ts#L69) and [here](https://github.com/midonet/tssrp6a/blob/ec90b27105b2b18cb7b57d469ffa2d5194bde19f/src/routines.ts#L110).

**Comment by bgrosse-midokura** [comment]
I think it is addressed in https://github.com/midonet/tssrp6a#recommendations

> The client can choose to exclude the identity of its computations or not. If excluded, the id cannot be changed.

But, I think this might be wrong, exactly the opposite: if **included**, then the ID cannot be changed.

The implementation would need to override `computeIdentityHash` as you pointed out, I'd guess just concatenating the the identity + password.

The _reason_ for why it's not done by default in the library? We matched the existing java "nimbus" implementation, so I guess there it is not included.

Not closing the issue, because I think now we should decide if the library should include identifier in default settings.

**Comment by JexSrs** [comment]
I understand, please keep an update here in case of any changes for this matter. Thank you.

Edit: In `computeIdentityHash` I can do as you said an concat it, what must change in [here](https://github.com/midonet/tssrp6a/blob/ec90b27105b2b18cb7b57d469ffa2d5194bde19f/src/routines.ts#L110)? Should I just add `bigIntToArrayBuffer(_I)` and `bigIntToArrayBuffer(_s)` at the start of the `hash` function?

**Comment by bgrosse-midokura** [comment]
Looking into that, meanwhile I can confirm that the reason we don't include identity is Nimbus: https://github.com/midonet/tssrp6a#notes

**Comment by JexSrs** [comment]
I understand what about this?

> In `computeIdentityHash` I can do as you said an concat it, what must change in [here](https://github.com/midonet/tssrp6a/blob/ec90b27105b2b18cb7b57d469ffa2d5194bde19f/src/routines.ts#L110)? Should I just add `bigIntToArrayBuffer(_I)` and `bigIntToArrayBuffer(_s)` at the start of the `hash` function?

Should I change only the `computeIdentityHash` or the `computeClientEvidence` too?

**Comment by bgrosse-midokura** [comment]
so I just tested: both will work, independently or together, as long as both client and server match the routine(s).

**Comment by JexSrs** [comment]
Perfect, can we safely say that will be more secure if we add it to both of them?

**Comment by bgrosse-midokura** [comment]
```typescript
    class IdentityRoutines extends SRPRoutines { ‣ 1 reference
      public async computeIdentityHash(I: string, P: string): Promise<ArrayBuffer> { ‣ 3 references
        return await this.hash(stringToArrayBuffer(I), stringToArrayBuffer(P));
      }
      public async computeClientEvidence( ‣ 3 references
        I: string,
        s: bigint,
        A: bigint,
        B: bigint,
        S: bigint,
      ): Promise<bigint> {
        return arrayBufferToBigInt(
          await this.hash(
            stringToArrayBuffer(I),
            bigIntToArrayBuffer(s),
            bigIntToArrayBuffer(A),
            bigIntToArrayBuffer(B),
            bigIntToArrayBuffer(S),
          ),
        );
      }
    }
```

**Comment by bgrosse-midokura** [comment]
> Perfect, can we safely say that will be more secure if we add it to both of them?

I will open a PR to discuss this further, I guess defaulting to including the identity is a major breaking change.

**Comment by JexSrs** [comment]
I will wait for updates, you can close this issue if you see fit.

**Comment by bgrosse-midokura** [comment]
I just found this https://github.com/midonet/tssrp6a/blob/master/test/srp6a.test.ts#L14 :sweat_smile:

**Comment by bgrosse-midokura** [comment]
But I'm left wondering about `computeClientEvidence`...

**Comment by JexSrs** [comment]
I agree, I believe we cannot change it the same way using `extends` because it belongs in utils which is not a class. Anyway I solved it by forking the project and change it directly. Also change all the code from async to sync, async seems unnecessary in case where no actual async functions where called.

**Comment by bgrosse-midokura** [comment]
Async _is_ needed when using async crypto APIs, like browser's WebCrypto IIRC.

I prefer to have one async implementation instead of one async and one sync.

**Comment by bgrosse-midokura** [comment]
I just tested, nimbus fails if computeClientEvidence includes the identity and salt - but I'm not 100% if that is truly related to computeIdentityHash.

**Comment by JexSrs** [comment]
I see, well I do not need browser implementation so I discarted WebCrypto from the code. Yes I just tested that as yell (`computeClientEvidence`) and no use (not sure why tho), but fortunately the typescript works just fine by including it to both functions.

**Comment by wise1994** [comment]
hi, first of all thank you for all you do.

I think "can change the identity" is a bad design. We should assume that a well-designed system should not use things like the user's email as the identifier. Just imagine, currently my website uses user email as id, one day in the future I need to add phone number login option, I need to use another id to associate them. Likewise, if I need to provide users with a publicly accessible home page, like Twitter, I need to expose the user's private data in a URL, like this: mysite.com/**@***.com/post_id

In fact, changing the identity requires no less work than changing the verifier.

Going back to the question itself, I think we should release a major version, exporting two classes, one that includes identity in computation by default, and one that doesn't. At some point in the future, replace the old implementation with the new implementation. But I'm not sure if this is a good idea.
